sets/sets.go

```diff
@@ -0,0 +1,80 @@
+// Package sets is a minimal implementation of a generic set data structure.
+
+package sets
+
+import (
+	"fmt"
+	"sort"
+
+	"golang.org/x/exp/constraints"
+)
+
+// Set is a minimal set that takes only ordered types: any type that supports the
+// operators < <= >= >.
+type Set[T constraints.Ordered] struct {
+	items map[T]struct{}
+}
+
+// New returns an empty set with capacity size. The capacity will grow and shrink as a
+// stdlib map.
+func New[T constraints.Ordered](size int) *Set[T] {
+	return &Set[T]{items: make(map[T]struct{}, size)}
+}
+
+// From returns a set from elements.
+func From[T constraints.Ordered](elements ...T) *Set[T] {
+	s := New[T](len(elements))
+	for _, i := range elements {
+		s.items[i] = struct{}{}
+	}
+	return s
+}
+
+// String returns a string representation of s, ordered. This allows to simply pass a
+// sets.Set as parameter to a function that expects a fmt.Stringer interface and obtain
+// a comparable string.
+func (s *Set[T]) String() string {
+	return fmt.Sprint(s.OrderedList())
+}
+
+func (s *Set[T]) Size() int {
+	return len(s.items)
+}
+
+// OrderedList returns a slice of the elements of s, ordered.
+// TODO This can probably be replaced in Go 1.20 when a generics slice packages reaches
+// the stdlib.
+func (s *Set[T]) OrderedList() []T {
+	elements := make([]T, 0, len(s.items))
+	for e := range s.items {
+		elements = append(elements, e)
+	}
+	sort.Slice(elements, func(i, j int) bool {
+		return elements[i] < elements[j]
+	})
+	return elements
+}
+
+// Contains returns true if s contains item.
+func (s *Set[T]) Contains(item T) bool {
+	_, found := s.items[item]
```

Comment:
Idiomatic map lookup.
It can be confusing, see https://gobyexample.com/maps for details; notice that the lookup can be done assigning to one or to two variables (special case).